url49526af437_Maria_G_CS325_tgz-pJ8-assignments_farmer_FarmerImplJ8/src/minidraw/standard/StandardDrawing.java

```diff
@@ -0,0 +1,146 @@
+  }
+
+  // === Delegation methods for the DrawingChangeListeners
+
+  /**
+   * Adds a listener for this drawing.
+   */
+  public void addDrawingChangeListener(DrawingChangeListener listener) {
+    listenerHandler.addDrawingChangeListener(listener);
+  }
+  
+  /**
+   * Removes a listener from this drawing.
+   */
+  public void removeDrawingChangeListener(DrawingChangeListener listener) {
+    listenerHandler.removeDrawingChangeListener(listener);
+  }
+
+  /**
+   * Invalidates a rectangle and merges it with the
+   * existing damaged area.
+   * @see FigureChangeListener
+   */
+  public void figureInvalidated(FigureChangeEvent e) {
+    listenerHandler.
+      fireDrawingInvalidated( this, e.getInvalidatedRectangle() );
+  }
+ 
+  public void figureChanged(FigureChangeEvent e){
+    listenerHandler.fireDrawingRequestUpdate( this );
+  }
+
+  public void requestUpdate() {
+    listenerHandler.fireDrawingRequestUpdate( this );
+  }
+
+  // === Delegation methods for the Selection handling
+
+  /**
+   * Get a list of all selected figures
+   */
+  public List<Figure> selection() {
+    return selectionHandler.selection();
+  }
+
+  /**
+   * Adds a figure to the current selection.
+   */
+  public void addToSelection(Figure figure) {
+    selectionHandler.addToSelection(figure);
+  }
+
+  /**
+   * Removes a figure from the selection.
+   */
+  public void removeFromSelection(Figure figure) {
+    selectionHandler.removeFromSelection(figure);
+  }
+
+  /**
+   * If a figure isn't selected it is added to the selection.
+   * Otherwise it is removed from the selection.
+   */
+  public void toggleSelection(Figure figure) {
+    selectionHandler.toggleSelection(figure);
+  }
+  
+  /**
+   * Clears the current selection.
+   */
+  public void clearSelection() {
+    selectionHandler.clearSelection();
+  }
+
+  // === We do some opportunistic reuse of the CompositeFigure but
+  // unfortunately it defines methods that are not usable for a
+  // Drawing.
+
+  /** should not be used by a drawing - the reason this method is
+      defined is because of the opportunistic reuse opportunities of
+      the CompositeFigure */
+  public void basicDisplayBox( int x, int y ) {
+    throw new 
+      RuntimeException( "StandardDrawing: basicDisplayBox should not be invoked");
+  }
+  
+  public Rectangle displayBox() {
+      throw new 
+        RuntimeException( "StandardDrawing: displayBox should not be invoked");
+  }
+
+
+  /**
+   * boolean that serves as a condition variable
+   * to lock the access to the drawing.
+   * The lock is recursive and we keep track of the current
+   * lock holder.
+   */
+  private transient Thread    fDrawingLockHolder = null;
+  
+  public synchronized void lock() {
+    // recursive lock
+    Thread current = Thread.currentThread();
+    if (fDrawingLockHolder == current) {
+      return;
+    }
+    while (fDrawingLockHolder != null) {
+      try { wait(); } catch (InterruptedException ex) { }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)